infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/StreamsSnapshotWriter.java

```diff
@@ -354,14 +334,77 @@ private void applyShadowStream(UUID streamId, long snapshot) {
+     * @return UUID of the original stream's shadow stream
+     */
+    private UUID getShadowStreamUUID(UUID regularStreamId) {
+        return CorfuRuntime.getStreamID(getFullyQualifiedTableName(CORFU_SYSTEM_NAMESPACE,
+                getShadowStreamName(regularStreamId)));
+    }
+
+    /**
+     * Rebuild the stream to tags map from registry table. This method will be invoked after TRANSFER
+     * phase of snapshot sync and before applying shadow streams to their original streams.
+     */
+    private void rebuildStreamToTagsMap() {
+        CorfuTable<TableName, CorfuRecord<TableDescriptors, TableMetadata>> registryTable =
+                rt.getTableRegistry().getRegistryTable();
+
+        registryTable.forEach((tableName, tableRecord) -> {
+            UUID streamId = CorfuRuntime.getStreamID(getFullyQualifiedTableName(tableName));
+            dataStreamToTagsMap.putIfAbsent(streamId, new HashSet<>());
+            dataStreamToTagsMap.get(streamId).addAll(
+                    tableRecord.getMetadata()
+                            .getTableOptions()
+                            .getStreamTagList()
+                            .stream()
+                            .map(streamTag -> TableRegistry.getStreamIdForStreamTag(tableName.getNamespace(), streamTag))
+                            .collect(Collectors.toList()));
+        });
+        // Get streams to replicate from registry table and update log replication config.
+        Set<UUID> streamsToReplicate = registryTable.entryStream()
+                .filter(entry -> entry.getValue().getMetadata().getTableOptions().getIsFederated())
+                .map(entry -> CorfuRuntime.getStreamID(getFullyQualifiedTableName(entry.getKey())))
+                .collect(Collectors.toSet());
+        // Update the dataStreamToTagsMap in LogReplicationConfig, clear the existing mapping as
+        // this is the point that dataStreamToTagsMap should be initialized in LogReplicationConfig
+        config.updateDataStreamToTagsMap(dataStreamToTagsMap, true);
+        config.getStreamsInfo().updateStreamIdsOnStandby(streamsToReplicate, true);
+        config.getStreamsInfo().updateStreamIdsOnStandby(regularToShadowStreamId.keySet(), false);
+    }
+
     /**
      * Read from shadowStream and append/apply to the actual stream
      */
     public void applyShadowStreams() {
         long snapshot = rt.getAddressSpaceView().getLogTail();
-        log.debug("Apply Shadow Streams, total={}", streamViewMap.size());
-        for (UUID regularStreamId : streamViewMap.keySet()) {
-            applyShadowStream(regularStreamId, snapshot);
+        log.debug("Apply Shadow Streams, total={}", regularToShadowStreamId.size());
+
+        for (UUID mergeOnlyStreamId : mergeOnlyStreams) {
+            applyShadowStream(mergeOnlyStreamId, snapshot);
+        }
+
+        // Rebuild the stream to tags map from registryTable
+        rebuildStreamToTagsMap();
```

Comment:
this function updates the stream ids to replicate in addition to the tags map.  Can you rename it to something like buildStreamsToReplicateAndTagsMap() ?